### PR TITLE
DOCSP-32655-EJSON-deserialize

### PR DESCRIPTION
**Pull request by jason-price-mongodb**
## DESCRIPTION

Document EJSON deserialize method.

## STAGING

https://preview-mongodbjasonpricemongodb.gatsbyjs.io/mongodb-shell/DOCSP-32655-EJSON-deserialize/reference/ejson/deserialize/

https://preview-mongodbjasonpricemongodb.gatsbyjs.io/mongodb-shell/DOCSP-32655-EJSON-deserialize/reference/ejson/

## JIRA

https://jira.mongodb.org/browse/DOCSP-32655

## BUILD LOG

https://workerpool-boxgs.mongodbstitch.com/pages/job.html?collName=queue&jobId=653306cf3ede3ec111b8a0c8

## Self-Review Checklist

- [ ] Is this free of any warnings or errors in the RST?
- [ ] Is this free of spelling errors?
- [ ] Is this free of grammatical errors?
- [ ] Is this free of staging / rendering issues?
- [ ] Are all the links working?

## External Review Requirements

[What's expected of an external reviewer?](https://wiki.corp.mongodb.com/display/DE/Reviewing+Guidelines+for+the+MongoDB+Server+Documentation)

### CODE REVIEW

**Comment by jason-price-mongodb** [comment]
@addaleax please review. Thanks.

File: source/reference/ejson.txt [hunk 1]
```diff
@@ -24,10 +24,15 @@ to transform BSON data.
 
 .. list-table::
    :header-rows: 1
+   :widths: 30, 70
 
    * - EJSON Method
      - Use
 
+   * - :ref:`EJSON.deserialize() <mongosh-ejson-deserialize>`
+     - Convert Extended JSON objects to JavaScript objects. This method
```
Comment (addaleax) [inline source/reference/ejson.txt:33]:
```suggestion
     - Convert Extended JSON objects to BSON objects. This method
```

We refer to these as `BSON objects` in the descriptions of the other three methods, so I think it’s worth being consistent here

File: source/reference/ejson/deserialize.txt [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+.. include:: /includes/links-urls/ejson.rst
+
+The ``EJSON.deserialize()`` method converts Extended JSON objects to
+JavaScript objects.
```
Comment (addaleax) [inline source/reference/ejson/deserialize.txt:24]:
```suggestion
BSON objects.
```

(same as above)

File: source/reference/ejson/deserialize.txt [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+ 
+        .. list-table::
+           :header-rows: 1
+           :widths: 30 70
+
+           * - Boolean Value
+             - Description
+
+           * - ``true``
+             - Return JSON object types. Default is ``true``.
```
Comment (addaleax) [inline source/reference/ejson/deserialize.txt:68]:
```suggestion
             - Return plain JavaScript object types. Default is ``true``.
```

The return value isn’t JSON, so I’m not sure if “JSON object types” really makes the most sense here.

File: source/reference/ejson/deserialize.txt [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+Interactive Mongo Shell EJSON.deserialize() Example
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following example retrieves the ``sales`` documents as an array and
+stores the results in the ``salesCollection`` object:
+
+.. code-block:: javascript
+
+   salesCollection = EJSON.deserialize( db.sales.find().toArray() )
```
Comment (addaleax) [inline source/reference/ejson/deserialize.txt:118]: Similar to `EJSON.parse()`, the use cases revolve around consuming external data, processing it, and then being able to interact with them like database contents.

Calling `EJSON.deserialize()` on objects coming from the database doesn’t make sense, the database already returns BSON objects, so there’s no need to convert them “back” to BSON objects.

The example below (reading data from disk and inserting it into the database) is definitely a better one.

File: source/reference/ejson/deserialize.txt [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+named ``salesFromDeserializeFile``, exit ``mongosh`` and then run this
+example from the command line:
+
+.. code-block:: sh
+
+   # Note: The example is formatted to fit the page.
+
+   mongosh --quiet  --eval "db.salesFromDeserializeFile.insertMany( \
+      EJSON.deserialize( JSON.parse ( \
+      fs.readFileSync( 'salesDeserialize.json', 'utf8' ) ) ) )"
```
Comment (addaleax) [inline source/reference/ejson/deserialize.txt:181]: It’s probably worth mentioning that, for this specific example, `EJSON.parse()` would be the more common approach, and that manually splitting it up into `JSON.parse()` + `EJSON.deserialize()` just gives the user a bit more flexibility.

File: source/reference/ejson/deserialize.txt [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+In the example:
+
+- ``fs.readFileSync()`` reads the ``salesDeserialize.json`` file and
+  interprets the content as ``utf8`` Unicode character strings.
+
+- ``JSON.parse()`` converts the strings read from the file to JSON.
+
+- ``EJSON.deserialize()`` outputs JSON pairs.
```
Comment (addaleax) [inline source/reference/ejson/deserialize.txt:190]: What’s a JSON pair?

**Review by jason-price-mongodb — COMMENTED** [review]
@addaleax please re-review. Thanks.

**Review by addaleax — APPROVED** [review]
Looks great!

**Pull request merged**